### ISSUE TITLE
Proposal: no-reject-literal rule

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#7685]
Continuing from https://github.com/eslint/eslint/issues/4092

**Please describe what the rule should do:**

This rule should attempt to prevent users from rejecting a Promise with a literal.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
Promise.reject(5);

Promise.reject("foo");

new Promise((resolve, reject) => {
  reject(5);
});
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Rejecting a Promise with a non-error is a bad practice, and is generally applicable regardless of what the Promises are being used for. We already have the [`no-throw-literal`](http://eslint.org/docs/rules/no-throw-literal) rule, which enforces the same thing with synchronous errors. This rule would complement `no-throw-literal` for async errors (i.e. Promise rejections).

Some Promise libraries, such as [bluebird](http://bluebirdjs.com/docs/warning-explanations.html#warning-a-promise-was-rejected-with-a-non-error), enforce this with a warning at runtime.

**Comment by TheSavior** [comment]
Cross posting this image to further demonstrate the issue.

![screen shot 2016-12-01 at 2 06 14 pm](https://cloud.githubusercontent.com/assets/249164/20818018/f08a5064-b7df-11e6-9fd3-67e70addf998.png)

**Comment by kaicataldo** [comment]
> Disallow object and array expressions as rejection reasons, in line with @platinumazure's suggestion earlier. Again, it seems like no-throw-literal only has that for backwards compatibility (or maybe it's just a bug), but I think we may as well make the behavior more robust.

Sorry, I'm having trouble finding this comment! Mind pointing me to it or explaining what the suggestion is?

**Comment by not-an-aardvark** [comment]
Sure, see [here](https://github.com/eslint/eslint/pull/7689#discussion_r90779675) and also the last section of the original PR description.

Basically, I think reporting object and array expressions is probably more correct, but it would be inconsistent with `no-throw-literal`. Since `no-throw-literal` has a misleading name and is only inconsistent for legacy reasons, I think it might be better to give the new rule a clearer name and behavior.
